lib/hobbes/eval/probes.C

```diff
@@ -0,0 +1,270 @@
+
+#include <hobbes/db/bindings.H>
+#include <hobbes/lang/tyunqualify.H>
+#include <hobbes/eval/probes.H>
+#include <hobbes/eval/cc.H>
+#include <hobbes/eval/jitcc.H>
+
+#include <atomic>
+
+namespace hobbes {
+
+static std::atomic<uint32_t> probeIdCtr{0};
+static uint64_t freshProbeId() noexcept {
+  return InjectedProbeMetadata::typeTag | static_cast<uint64_t>(probeIdCtr++);
+}
+
+struct injectProbeF : public op {
+  llvm::Value* apply(jitcc* c, const MonoTypes& tys, const MonoTypePtr& /*rty*/, const Exprs& es) {
+    // Requires us to be able to fully evaluate the argument to the compile-time string
+    MkArray* probeNameArr = is<MkArray>(stripAssumpHead(es[0]));
+    if (!probeNameArr) {
+      throw annotated_error(*es[0], "injectProbe intrinsic expected compile-time [char] as argument" + show(tys[0]));
+    }
+    if (probeNameArr->values().size() == 0) {
+      throw annotated_error(*es[0], "injectProbe intrinsic expected compile-time [char] as argument" + show(tys[0]));
+    }
+    std::string probeName = "";
+    for (auto v : probeNameArr->values()) {
+      Char* casted = is<Char>(v);
+      if (!casted) {
+        throw annotated_error(*es[0], "injectProbe intrinsic expected compile-time [char] as argument" + show(tys[0]));
+      }
+      probeName += casted->value();
+    }
+
+    const uint64_t probeId = freshProbeId();
+    constexpr uint32_t caveSize = 5; // Size in bytes of the nop to insert at patchpoint. Xpedite expects 5
+
+    // we need to register the probe (id and name) with the compiler such that the stackmap
+    // can be walked and probe location extracted & dumped to disk for xpedite.
+    c->addProbeMapping(
+      InjectedProbeMetadata{
+        probeId,
+        probeName,
+        es[0]->la().filename(),
+        es[0]->la().p0.first,  // lineno
+        "N/A",  // function name
+        reinterpret_cast<void*>(0), // callsite - to be filled later
+        reinterpret_cast<void*>(0), // returnsite - to be filled later
+      }
+    );
+
+    // patchpoints are utterly broken in patchpoint followed by call scenarios under x86_64
+    // instead we use stackmap followed by a manual inline asm 5 byte nop
+    llvm::InlineAsm* multibyteNop = llvm::InlineAsm::get(
+      llvm::FunctionType::get(c->builder()->getVoidTy(), false /* isVarArg */),
+      "nopl 0(%eax, %eax, 1)", // 5 byte nop
```

Comment:
Yeah, now we're tied to x86_64 anyway.

Reply:
This is non-portable. It defeats the purpose of having a portable backend generator such LLVM. In 10 years from now, when I want to port Hobbes on my Haskus hand-held ARM device, what kind of pain will this cause me ?

Reply:
Maybe it's something that should be kept in the back-end library anyway so that when you do have the job of moving to a different architecture then you have a known boundary to cover.

Realistically, x86_64 is our main use-case, but we should factor this kind of thing into a platform-specific backend IMHO.

Reply:
Worst case scenario: You just replace it with N inline no-op instructions (You'd be hard pressed to find an common architecture without them). LLVM Stackmaps don't even work on most platforms (I believe x86_64, ppc and Aarch64) as it is, so you have have bigger things to worry about because it wouldn't even be reasonable to obtain the locations of the caves to begin with. Having said all that, there's little reason this couldn't be abstracted behind a simple platform-specific C++ function call.